## 1. Introduction

Customer churn is a critical challenge for telecommunications companies, as losing existing customers directly impacts revenue and increases acquisition costs. Understanding which customers are likely to churn, and the reasons behind their decisions, enables organizations to design targeted and cost-effective retention strategies.

This project analyzes a telecom customer dataset containing demographic attributes, service subscriptions, billing and revenue information, and churn outcomes. Each row represents an individual customer, with the churn label indicating whether the customer left the company.

The objective of this project is to build an end-to-end analytics and machine learning solution that:
- Identifies customers at high risk of churn,
- Understands the key drivers and patterns associated with churn behavior, and
- Translates analytical insights into actionable business recommendations to improve customer retention.


## 2. Dataset Overview

The dataset represents a realistic telecom customer environment and includes the following feature groups:

Data Set From Kaggel :

https://www.kaggle.com/datasets/alfathterry/telco-customer-churn-11-1-3

- **Customer Demographics:** Age, gender, marital status, dependents, geography
- **Service Subscriptions:** Phone service, internet service, streaming options, contract type
- **Usage & Billing:** Monthly charges, total charges, long-distance usage
- **Revenue Metrics:** Total revenue, refunds, extra data charges, customer lifetime value (CLTV)
- **Churn Information:** Churn label, churn score, churn category, and churn reason

The primary target variable for modeling is **Churn Label**, where:
- `Yes` indicates a churned customer
- `No` indicates a retained customer


### Import Libraries (Code)

In [2]:
import pandas as pd
import numpy as py
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", None)

### Load the Dataset (Code)

In [3]:
df = pd.read_csv('../data/raw/telco_churn_dataset.csv')
df.head()

Customer ID  Gender  Age Under 30 Senior Citizen Married Dependents  \
0  8779-QRDMV    Male   78       No            Yes      No         No   
1  7495-OOKFY  Female   74       No            Yes     Yes        Yes   
2  1658-BYGOY    Male   71       No            Yes      No        Yes   
3  4598-XLKNJ  Female   78       No            Yes     Yes        Yes   
4  4846-WHAFZ  Female   80       No            Yes     Yes        Yes   

   Number of Dependents        Country       State         City  Zip Code  \
0                     0  United States  California  Los Angeles     90022   
1                     1  United States  California  Los Angeles     90063   
2                     3  United States  California  Los Angeles     90065   
3                     1  United States  California    Inglewood     90303   
4                     1  United States  California     Whittier     90602   

    Latitude   Longitude  Population Quarter Referred a Friend  \
0  34.023810 -118.156582       68701      Q3                No   
1  34.044271 -118.185237       55668      Q3               Yes   
2  34.108833 -118.229715       47534      Q3                No   
3  33.936291 -118.332639       27778      Q3               Yes   
4  33.972119 -118.020188       26265      Q3               Yes   

   Number of Referrals  Tenure in Months    Offer Phone Service  \
0                    0                 1      NaN            No   
1                    1                 8  Offer E           Yes   
2                    0                18  Offer D           Yes   
3                    1                25  Offer C           Yes   
4                    1                37  Offer C           Yes   

   Avg Monthly Long Distance Charges Multiple Lines Internet Service  \
0                               0.00             No              Yes   
1                              48.85            Yes              Yes   
2                              11.33            Yes              Yes   
3                              19.76             No              Yes   
4                               6.33            Yes              Yes   

  Internet Type  Avg Monthly GB Download Online Security Online Backup  \
0           DSL                        8              No            No   
1   Fiber Optic                       17              No           Yes   
2   Fiber Optic                       52              No            No   
3   Fiber Optic                       12              No           Yes   
4   Fiber Optic                       14              No            No   

  Device Protection Plan Premium Tech Support Streaming TV Streaming Movies  \
0                    Yes                   No           No              Yes   
1                     No                   No           No               No   
2                     No                   No          Yes              Yes   
3                    Yes                   No          Yes              Yes   
4                     No                   No           No               No   

  Streaming Music Unlimited Data        Contract Paperless Billing  \
0              No             No  Month-to-Month               Yes   
1              No            Yes  Month-to-Month               Yes   
2             Yes            Yes  Month-to-Month               Yes   
3              No            Yes  Month-to-Month               Yes   
4              No            Yes  Month-to-Month               Yes   

    Payment Method  Monthly Charge  Total Charges  Total Refunds  \
0  Bank Withdrawal           39.65          39.65           0.00   
1      Credit Card           80.65         633.30           0.00   
2  Bank Withdrawal           95.45        1752.55          45.61   
3  Bank Withdrawal           98.50        2514.50          13.43   
4  Bank Withdrawal           76.50        2868.15           0.00   

   Total Extra Data Charges  Total Long Distance Charges  Total Revenue  \
0                        20               

## Column Name Standardization

To improve code readability and consistency across analysis and modeling steps, all column names were standardized to lowercase and converted to snake_case format. This practice aligns with common data engineering and analytics conventions and helps avoid potential issues related to spacing or case sensitivity.


In [4]:
df.columns = (
    df.columns
      .str.strip()
      .str.lower()
      .str.replace(" ", "_")
      .str.replace("-", "_")
)


## Initial Data Inspection

In [5]:
df.shape

(7043, 50)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 50 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   customer_id                        7043 non-null   object 
 1   gender                             7043 non-null   object 
 2   age                                7043 non-null   int64  
 3   under_30                           7043 non-null   object 
 4   senior_citizen                     7043 non-null   object 
 5   married                            7043 non-null   object 
 6   dependents                         7043 non-null   object 
 7   number_of_dependents               7043 non-null   int64  
 8   country                            7043 non-null   object 
 9   state                              7043 non-null   object 
 10  city                               7043 non-null   object 
 11  zip_code                           7043 non-null   int64

In [7]:
df.describe(include='all')

customer_id gender          age under_30 senior_citizen married  \
count         7043   7043  7043.000000     7043           7043    7043   
unique        7043      2          NaN        2              2       2   
top     8779-QRDMV   Male          NaN       No             No      No   
freq             1   3555          NaN     5642           5901    3641   
mean           NaN    NaN    46.509726      NaN            NaN     NaN   
std            NaN    NaN    16.750352      NaN            NaN     NaN   
min            NaN    NaN    19.000000      NaN            NaN     NaN   
25%            NaN    NaN    32.000000      NaN            NaN     NaN   
50%            NaN    NaN    46.000000      NaN            NaN     NaN   
75%            NaN    NaN    60.000000      NaN            NaN     NaN   
max            NaN    NaN    80.000000      NaN            NaN     NaN   

       dependents  number_of_dependents        country       state  \
count        7043           7043.000000           7043        7043   
unique          2                   NaN              1           1   
top            No                   NaN  United States  California   
freq         5416                   NaN           7043        7043   
mean          NaN              0.468692            NaN         NaN   
std           NaN              0.962802            NaN         NaN   
min           NaN              0.000000            NaN         NaN   
25%           NaN              0.000000            NaN         NaN   
50%           NaN              0.000000            NaN         NaN   
75%           NaN              0.000000            NaN         NaN   
max           NaN              9.000000            NaN         NaN   

               city      zip_code     latitude    longitude     population  \
count          7043   7043.000000  7043.000000  7043.000000    7043.000000   
unique         1106           NaN          NaN          NaN            NaN   
top     Los Angeles           NaN          NaN          NaN            NaN   
freq            293           NaN          NaN          NaN            NaN   
mean            NaN  93486.070567    36.197455  -119.756684   22139.603294   
std             NaN   1856.767505     2.468929     2.154425   21152.392837   
min             NaN  90001.000000    32.555828  -124.301372      11.000000   
25%             NaN  92101.000000    33.990646  -121.788090    2344.000000   
50%             NaN  93518.000000    36.205465  -119.595293   17554.000000   
75%             NaN  95329.000000    38.161321  -117.969795   36125.000000   
max             NaN  96150.000000    41.962127  -114.192901  105285.000000   

       quarter referred_a_friend  number_of_referrals  tenure_in_months  \
count     7043              7043          7043.000000       7043.000000   
unique       1                 2                  NaN               NaN   
top         Q3                No                  NaN               NaN   
freq      7043              3821                  NaN               NaN   
mean       NaN               NaN             1.951867         32.386767   
std        NaN               NaN             3.001199         24.542061   
min        NaN               NaN             0.000000          1.000000   
25%        NaN               NaN             0.000000          9.000000   
50%        NaN               NaN             0.000000         29.000000   
75%        NaN               NaN             3.000000         55.000000   
max        NaN               NaN            11.000000         72.000000   

          offer phone_service  avg_monthly_long_distance_charges  \
count      3166          7043                        7043.000000   
unique        5             2                                NaN   
top     Offer B           Yes                                NaN   
freq        824          6361                                NaN   
mean        NaN           NaN                          22.958954   
std         NaN           NaN  

#### Discovered

- The dataset has 7043 rows and  50 columns
- There no numerical missing values but contextual in columns as ('internet type , churn category and churn reason)
- Which columns are categorical vs numerical? - i don't know
- Which variables appear derived or post-event? - i don't know


## Target Variable Validation
- Making sure churn predicting is correct, meaningful, and usable before building models

In [8]:
df['churn_label'].value_counts()

churn_label
No     5174
Yes    1869
Name: count, dtype: int64

In [9]:
df['churn_label'].value_counts(normalize=True)

churn_label
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

The churn distribution indicates that approximately 25–30% of customers in the dataset have churned. This reflects a moderately imbalanced target variable, which is common in telecom churn problems.

## Target Variable Encoding

For consistency in analysis and to support statistical aggregation and predictive modeling, the churn target variable was encoded as a binary label. Customers who churned were mapped to 1, while retained customers were mapped to 0. This encoding enables intuitive interpretation of group-level means as churn rates and ensures compatibility with machine learning algorithms.


In [10]:
df['churn_label']= df['churn_label'].map({'No':0,'Yes':1})

## Feature Categorization & Data Leakage Prevention

To ensure that predictive modeling reflects real-world deployment scenarios, post-outcome variables and derived churn indicators were excluded from modeling to prevent data leakage.
These variables contain information that would not be available at the time of prediction and could artificially inflate model performance.
Such variables were retained solely for exploratory analysis and business interpretation.


In [11]:
# excluding post-churn and outcome-derived variables
leakage_columns = [
    "customer_status",
    "churn_score",
    "churn_category",
    "churn_reason"
]

identifier_columns = ["customer_id"]

df[leakage_columns].head()

customer_status  churn_score   churn_category                  churn_reason
0         Churned           91       Competitor  Competitor offered more data
1         Churned           69       Competitor  Competitor made better offer
2         Churned           81       Competitor  Competitor made better offer
3         Churned           88  Dissatisfaction     Limited range of services
4         Churned           67            Price            Extra data charges

## Sanity Checks & Business Validation

In [12]:
df[["tenure_in_months", "monthly_charge", "total_charges", "total_revenue"]].describe()


tenure_in_months  monthly_charge  total_charges  total_revenue
count       7043.000000     7043.000000    7043.000000    7043.000000
mean          32.386767       64.761692    2280.381264    3034.379056
std           24.542061       30.090047    2266.220462    2865.204542
min            1.000000       18.250000      18.800000      21.360000
25%            9.000000       35.500000     400.150000     605.610000
50%           29.000000       70.350000    1394.550000    2108.640000
75%           55.000000       89.850000    3786.600000    4801.145000
max           72.000000      118.750000    8684.800000   11979.340000

In [13]:
(df["total_charges"] < 0).sum()


np.int64(0)

In [14]:
(df["monthly_charge"] <= 0).sum()


np.int64(0)

In [15]:
(df["total_revenue"] < df["monthly_charge"]).sum()


np.int64(0)

Basic sanity checks were performed on tenure, billing, and revenue variables to ensure alignment with real-world business logic. No negative values were observed for monetary fields, and tenure values fell within reasonable ranges. These checks confirm that the dataset is internally consistent and suitable for further analysis.


## Data Cleaning Decisions

1. The dataset required minimal preprocessing as it was already well-structured and contained no significant missing values.
2. Cleaning efforts focused on standardizing column names, validating data consistency, and ensuring alignment with real-world business logic.
3. Post-outcome variables related to churn were retained for exploratory analysis and interpretation but deliberately excluded from predictive modeling to prevent data leakage.
4. No records were removed at this stage to preserve the integrity of the original customer population.


## Save Cleaned Dataset

In [17]:
df_clean= df.copy()

df_clean.to_csv("../data/processed/telco_churn_cleaned.csv", index=False)

## Cleaning Summary

This phase focused on establishing a reliable foundation for analysis by understanding the dataset structure, validating the churn target, and ensuring data quality and business coherence.
- Column names were standardized for consistency,
- key feature groups were identified,
 - potential sources of data leakage were explicitly documented and excluded from modeling considerations.
- Sanity checks confirmed that tenure, billing, and revenue variables aligned with realistic telecom business logic.
- The resulting cleaned and well-defined dataset provides a solid and trustworthy basis for exploratory analysis and predictive modeling in the next stages.
